libopenage/gamestate/game_main.cpp

```diff
@@ -22,28 +22,32 @@ GameMain::GameMain(const Generator &generator)
 	// players
 	unsigned int i = 0;
 	for (auto &name : generator.player_names()) {
-		this->players.emplace_back(this->add_civ(i), i, name);
+		this->players.push_back(new Player(this->add_civ(i), i, name));
```

Comment:
The `shared_ptr`/`unique_ptr` actually removes comlexity as you don't need to delete the players in the deconstructor then. And it is exception-safe only then.
The `vector` uses the heap internally but with the `new Player` the vector stores the pointer list on the heap which point to Players allocated somewhere else.

We have two choices:

store the player in a `shared_ptr`
```
    this->players.reserve(generator.player_names().size());
    for (auto &name : generator.player_names()) {
        this->players.push_back(std::make_shared<Player>(this->add_civ(i), i, name));
    }
```
or store the player data directly in the vector
```
    this->players.reserve(generator.player_names().size());
    for (auto &name : generator.player_names()) {
        this->players.push_back(Player{this->add_civ(i), i, name});
    }
```

The `shared_ptr` storage has the advantage that when the vector invalidates its iterators/pointers because of resizing, the raw pointers to players would remain the same.
But nobody needs to hold raw pointers to `Player`s, then the direct storage is better (and more cache-friendly).